core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -124,7 +124,7 @@ static TableMetadata newTableMetadata(
     MetricsConfig.fromProperties(properties).validateReferencedColumns(schema);
 
     return new Builder()
-        .upgradeFormatVersion(formatVersion)
+        .setFormatVersion(formatVersion)
```

Comment:
I think it's a bit dangerous to expose the ability to set rather than upgrade the format version. I think people are going to call it when they shouldn't. Is it actually needed here?

Reply:
The builder inits the format version to the default format version (2 now) and if users explicitly pass 1 as the desired format version, create table statements start to fail as it is not allowed to downgrade.

We can make this package-private given that it is only used during table creation. Or we could lazily init the default format version. Either way would work for me.

Reply:
I made it private, which is still accessible in `TableMetadata`. It is kind of weird to call `upgrade` when creating a fresh metadata file.

Reply:
I feel like `upgradeFormatVersion` (despite the name) might be more consistent the other setters, which perform checks and add the MetadataUpdate as needed.

Reply:
I don't think the validation in `upgradeFormatVersion` is applicable when creating new tables. I also don't think we are supposed to generate an update cause we are not updating the format version, the table does not exist.